### PR TITLE
[RFC] Switch from black to ruff fmt

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124845

Note: need to install lintrunner_adapters for CI, haven't done it yet.

See discussion at https://github.com/pytorch/pytorch/issues/123062

On my Macbook M1 Pro, it took ruff fmt 17s to generate this diff, and it took ufmt 1m9s to undo the changes from ruff fmt. So that's about a 6x speedup.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @albanD @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC

### CODE REVIEW

**Comment by ezyang** [comment]
What I am looking for out of the RFC: do you like the format change? Hate it? Are you willing to deal with 2k lines of churn just to get 6x faster format speed?

**Comment by janeyx99** [comment]
Seems worth it for the speedup; I looked over the optim files and both formats look equally acceptable, so…I vote yes.

**Comment by huydhn** [comment]
The speedup is nice and those files w.r.t CI looks good to me.  However, there might be some legwork needed to roll this out because I think ruff fmt seems to fumble when it comes to:

1. MYPY `# type: ignore` in some cases when the directives are moved to a different line.
2. It doesn't seem to honor `# fmt: skip` from what I see.

Also do we want to lower the line length limit from 120 to 88 as part of this change?

**Comment by albanD** [comment]
> Also do we want to lower the line length limit from 120 to 88 as part of this change?

Do we have a good reason to do that? I feel like most screens can show 120 these days :)
And 120 already make for some crazy line breaks, I would prefer to keep it at 120 personally.

I'm happy with the move for the speedup. I think we've been doing so many lint (some are speedup/bugfix!) enablement that the churn in 2k LOC is not much in the grand scheme of things.

**Comment by Skylion007** [comment]
Can we enable the ruff isort rules at the same time?

**Comment by Skylion007** [comment]
> The speedup is nice and those files w.r.t CI looks good to me. However, there might be some legwork needed to roll this out because I think ruff fmt seems to fumble when it comes to:
> 
>     1. MYPY `# type: ignore` in some cases where the directives are moved to a different line and failed.
> 
>     2. It doesn't seem to honor `# fmt: skip` from what I see.
> 
> 
> Also do we want to lower the line length limit from 120 to 88 as part of this change? (as I see it's currently part of the PR)

It should honor `#fmt: skip` https://docs.astral.sh/ruff/formatter/#docstring-formatting

**Comment by ezyang** [comment]
> Also do we want to lower the line length limit from 120 to 88 as part of this change? (as I see it's currently part of the PR)

This is lowering it to match our current black settings, which are 88. If we revert that change, we have substantially more changes in the diff.

> MYPY # type: ignore in some cases where the directives are moved to a different line and failed.

Black does a really bad job at this too, probably number one problem we face when turning on autoformatting.

> Can we enable the ruff isort rules at the same time?

Yes. Actually, we have an interesting problem here, which is that ufmt also can backend to ruff format, so we have to decide if we want to use ruff isort or usort.

**Comment by Skylion007** [comment]
ruff isort is just faster, any reason to use ufmt over it?

**Comment by amyreese** [comment]
> > MYPY # type: ignore in some cases where the directives are moved to a different line and failed.
> 
> Black does a really bad job at this too, probably number one problem we face when turning on autoformatting.

FWIW, Black 24.x does a much better job at keeping type ignores in the same place, even if it results in exceeding line lengths.

File: pyproject.toml [hunk 1]
```diff
@@ -40,7 +40,7 @@ standard_library = ["typing_extensions"]
 
 [tool.ruff]
 target-version = "py38"
-line-length = 120
+line-length = 88
```
Comment (Skylion007) [inline pyproject.toml:43]: Need to select the I001 to enable RUFF isort: https://docs.astral.sh/ruff/rules/unsorted-imports/

**Comment by Skylion007** [comment]
So the one potential use case for keeping ufmt for is now is that the ruff Isort implementation follows the linter exclude rules, not the fmt exclude rules for now: https://docs.astral.sh/ruff/formatter/#sorting-imports

**Comment by ezyang** [comment]
I don't have much of an opinion on import sorting haha, but I guess how exactly you setup the heuristics there, there is a lot of degrees of freedom.

**Comment by Skylion007** [comment]
Apparently the formatting handles isort formatting and you don't need to enable the ISC rules.

**Comment by XuehaiPan** [comment]
I'm glad to see we are switching to a unified style linter. Also, a faster linter allows us to enable `pre-commit` hook at committing before submitting the PR. This can alleviate our CI pressure.

But I'm concerned about how stable the `ruff formatter` style is.

Originally posted at https://github.com/pytorch/pytorch/pull/129870#issuecomment-2201164764:

> Refer to `black` documentation [How stable is Black’s style?](https://black.readthedocs.io/en/stable/faq.html#how-stable-is-black-s-style):
> 
> > Stable. Black aims to enforce one style and one style only, with some room for pragmatism. See [The Black Code Style](https://black.readthedocs.io/en/stable/the_black_code_style/index.html) for more details.
> > 
> > _**Starting in 2022, the formatting output is stable for the releases made in the same year (other than unintentional bugs).**_ At the beginning of every year, the first release will make changes to the stable style. It is possible to opt in to the latest formatting styles using the --preview flag.
> 
> `black 24.*.*` will produce a consistent 2024 style.

I haven't seen any disclaimer about the style stability on ruff documentation: [The Ruff Formatter](https://docs.astral.sh/ruff/formatter).

Another concern is we now have two linter runs `ruff` (`RUFF` and `RUFF_FORMAT`). Should we have to pin the `ruff` version in these two linter separately?

**Comment by XuehaiPan** [comment]
I opened a new diff for `black` 2024 style vs. `ruff format` (based on `ufmt ruff-api` (`ruff` 0.3.7)).

- #130029

**Comment by ezyang** [comment]
The Ruff docs seem to imply that they try to follow Black formatting as much as possible, with only some divergences.

**Pull request closed without merging**